delivery_auto_refresh/models/sale_order.py

```diff
@@ -8,115 +9,185 @@
+        if new_vals:
+            delivery_line.write(new_vals)
 
     def _auto_refresh_delivery(self):
         self.ensure_one()
-        # Make sure that if you have removed the carrier, the line is gone
-        if self.state in {"draft", "sent"}:
-            # Context added to avoid the recursive calls and save the new
-            # value of carrier_id
-            self.with_context(auto_refresh_delivery=True)._remove_delivery_line()
-        if self._get_param_auto_add_delivery_line() and self.carrier_id:
-            if self.state in {"draft", "sent"}:
-                price_unit = self.carrier_id.rate_shipment(self)["price"]
-                if not self.is_all_service:
-                    self._create_delivery_line(self.carrier_id, price_unit)
-                self.with_context(auto_refresh_delivery=True).write(
-                    {"recompute_delivery_price": False}
-                )
-
-    @api.model
-    def create(self, vals):
-        """Create or refresh delivery line on create."""
-        order = super().create(vals)
-        order._auto_refresh_delivery()
-        return order
+        if self.env.context.get("auto_refresh_delivery"):
+            return
+        if not self._is_auto_add_delivery_line():
+            return
+        if self.state not in ("draft", "sent"):
+            return
+
+        self = self.with_context(auto_refresh_delivery=True)
+
+        if not self.carrier_id:
+            self._set_delivery_carrier()
+
+        if not self.carrier_id:
+            self._remove_delivery_line()
+        elif self.is_all_service:
+            self._remove_delivery_line()
+        else:
+            price_unit = self.carrier_id.rate_shipment(self)["price"]
+            delivery_lines = self.order_line.filtered("is_delivery")
+            if not delivery_lines:
+                self._create_delivery_line(self.carrier_id, price_unit)
+            elif len(delivery_lines) > 1:
+                delivery_discount = self._get_delivery_discount(delivery_lines)
+                self._remove_delivery_line()
+                sol = self._create_delivery_line(self.carrier_id, price_unit)
+                if delivery_discount and sol:
+                    sol.discount = delivery_discount
+            else:
+                self._update_delivery_line(delivery_lines[0], price_unit)
+        if self.recompute_delivery_price:
+            self.recompute_delivery_price = False
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        # Prevent to refresh delivery in the call to super
+        orders = (
+            super(SaleOrder, self.with_context(auto_refresh_delivery=True))
+            .create(vals_list)
+            .with_context(auto_refresh_delivery=False)
+        )
+        for order in orders:
+            # Migration Note 17.0: move this to module sale_order_carrier_auto_assign
+            if not order.carrier_id and order._is_auto_set_carrier_on_create():
+                order._set_delivery_carrier()
```

Comment:
Why doing this instead of converting the field into a computed writable field that makes the code more declarative?